File: 275.h-指数-ii.py
#
# @lc app=leetcode.cn id=275 lang=python3
#
# [275] H 指数 II
#

# @lc code=start
class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        left, right = 0, n - 1
        while left <= right:
            mid = (right - left) // 2 + left
            diff = n - mid
            if citations[mid] == diff:
                return diff
            if citations[mid] < diff:
                left = mid + 1
            else:
                right = mid - 1
        return n - left     # if all citations are greater than n

# @lc code=end

File: 275.h-指数-ii.java
/*
 * @lc app=leetcode.cn id=275 lang=java
 *
 * [275] H 指数 II
 */

// @lc code=start
class Solution {
    public int hIndex(int[] citations) {
        // Binary search is the most efficient when data is sorted
        int n = citations.length, l = 0, r = n - 1;
        while (l <= r) {
            int m = ((r - l) >> 1) + l;
            int h = n - m;
            if (citations[m] == h) {
                return h;
            } else if (citations[m] > h) {
                r = m - 1;
            } else {
                l = m + 1;
            }
        }
        return n - l;
    }
}
// @lc code=end

